contrib/p4/p4execution.py

```diff
@@ -0,0 +1,838 @@
+        test_info = test_info.decode()
+
+        test_dict = self.extract_info_from_test_info(test_info)
+
+        return test_dict
+
+    def extract_info_from_test_info(self, test_info):
+        test_info = test_info.split("Test List:")[1]
+        test_modules = test_info.split("Module ")
+        nr_of_modules = len(test_modules) - 1
+        test_modules[len(test_modules) - 1] = test_modules[len(test_modules) - 1].split(
+            "\n{0}".format(nr_of_modules)
+        )[0]
+
+        test_dict = {}
+
+        for i in range(nr_of_modules):
+            test = test_modules[i + 1].split("\n")
+            module_name = test.pop(0).split(":")[0]
+
+            test_names = []
+            for test_string in test:
+                if not str.isspace(test_string) and test_string:
+                    test_names.append(test_string.split(":")[0].strip())
+
+            test_dict[module_name] = test_names
+
+        return test_dict
+
+    def get_system_info(self):
+        return systeminfo.SystemInfo()
+
+    def read_folder_paths(self, benchmark):
+
+        switch_folder = ""
+        ptf_folder = ""
+        network_config = ""
+        option_index = 0
+
+        while option_index < len(benchmark.options):
+            try:
+                if "switch" in benchmark.options[option_index].lower():
+                    switch_folder = benchmark.options[option_index + 1]
+                elif "ptf" in benchmark.options[option_index].lower():
+                    ptf_folder = benchmark.options[option_index + 1]
+                elif "network_config" in benchmark.options[option_index].lower():
+                    network_config = benchmark.options[option_index + 1]
+            except:
+                self.close()
+                raise BenchExecException(
+                    benchmark.options[option_index]
+                    + " did not match any expected options"
+                )
+
+            option_index += 2
+
+        if "~" in switch_folder:
+            switch_folder = self.extract_path(switch_folder)
+        if "~" in ptf_folder:
+            ptf_folder = self.extract_path(ptf_folder)
+        if "~" in network_config:
+            network_config = self.extract_path(network_config)
+
+        return switch_folder, ptf_folder, network_config
+
+    def extract_path(self, path):
+        import subprocess
+
+        split = subprocess.run(["pwd"], capture_output=True).stdout.decode().split("/")
+        home_dir = f"/{split[1]}/{split[2]}"
+
+        new_path = path.replace("~", home_dir)
+
+        return new_path
+
+    def stop(self):
+        """
+        Needed for automatic cleanup for benchec.
+        """
+        self.close()
+
+    def close(self):
+        """
+        Cleans up all the running containers and clear all created namespaces. Should be called when test is done.
+        """
+        logging.info("Closing containers and cleaning up namespace")
+        container_threads = []
```

Comment:
I think it would be easier to use a `concurrent.futures.ThreadPoolExecutor` and simply submit cleanup jobs there instead of manual thread handling?